In [18]:
import pandas as pd
import numpy as np

In [19]:
file = 'training_sample.tsv'

In [20]:
column_names = ["text_tokens", "hashtags", "tweet_id", "present_media", "present_links", "present_domains",\
                "tweet_type", "language", "tweet_timestamp", "engaged_with_user_id", "engaged_with_user_follower_count",\
               "engaged_with_user_following_count", "engaged_with_user_is_verified", "engaged_with_user_account_creation",\
               "engaging_user_id", "engaging_user_follower_count", "engaging_user_following_count", "engaging_user_is_verified",\
               "engaging_user_account_creation", "engaged_follows_engaging", "reply_timestamp", "retweet_timestamp", "retweet_with_comment_timestamp", "like_timestamp"]

In [21]:
df = pd.read_csv(file, delimiter='\x01')

In [22]:
df.head()

101\t56898\t137\t174\t63247\t10526\t131\t3197\t8747\t8747\t6455\t3205\t14120\t131\t120\t120\t188\t119\t11170\t120\t176\t11090\t10305\t10686\t11211\t10116\t11127\t10237\t11779\t11517\t102  \
0  101\t102463\t10230\t10105\t21040\t10169\t12811...                                                                                                                                           
1  101\t56898\t137\t11255\t22037\t10263\t168\t111...                                                                                                                                           
2  101\t13073\t28757\t106\t100\t14120\t131\t120\t...                                                                                                                                           
3  101\t3460\t1923\t6632\t2824\t30368\t2179\t1881...                                                                                                                                           
4  101\t46242\t40751\t161\t100062\t10107\t10114\t...                                                                                                                                           

                         Unnamed: 1  3C21DCFB8E3FEC1CB3D2BFB413A78220  Video  \
0                               NaN  3D87CC3655C276F1771752081423B405    NaN   
1  DB32BD91C2F1B37BE700F374A07FBC61  3701848B96AA740528A2B0E247777D7D    NaN   
2                               NaN  18176C6AD2871729384062F073CCE94D  Video   
3                               NaN  AF11AF01F842E7F120667B7B0B38676D    NaN   
4                               NaN  4EC1C41302B820B2459337FE31859B16    NaN   

                         Unnamed: 4                        Unnamed: 5  \
0  BB422AA00380E45F312FD2CAA75F4960  92D397F8E0F1E77B36B8C612C2C51E23   
1  2423BA02A75DB2189335DDC3FB6B74A1  6D323BE93766E79BE423FAC5C28BE39B   
2                               NaN                               NaN   
3                               NaN                               NaN   
4  BB79CD318A68247B64F0E0BE7AFD5A92  3896E26D12C903F0A00B6B1BE9A9BEA3   

    Retweet  76B8A9C3013AE6414A3E6012413CDC3B  1581467323  \
0  TopLevel  D3164C7FBCF2565DDF915B1B3AEFB1DC  1580975391   
1   Retweet  22C448FF81263D4BAF2A176145EE9EAD  1581257232   
2  TopLevel  D3164C7FBCF2565DDF915B1B3AEFB1DC  1581164292   
3     Quote  22C448FF81263D4BAF2A176145EE9EAD  1581233650   
4  TopLevel  D3164C7FBCF2565DDF915B1B3AEFB1DC  1581426841   

   D1AA2C85FA644D64346EDD88470525F2       737   706  False  1403069820  \
0  4DC65AC7BD963DE1F7617C047C33DE99  52366425  2383   True  1230139136   
1  5C671539CB41B9807E209349B101E9FF       988   167  False  1530094483   
2  70B900BE17416923D1E236A38798F202   1228134  5413  False  1378699943   
3  E94C0E9E8494F3D603F9D1A5C5242E3D        73   299  False  1549054499   
4  959F94011E65010724E85D1B5CBAE18F     45488  6709   True  1415444172   

   000046C8606F1C3F5A7296222C88084B  131  2105  False.1  1573978269  False.2  \
0  00006047187D0D18598EF12A650E1DAC   22    50    False  1340673962    False   
1  0000648BAA193AE4C625DDF789B57172  251   719    False  1456473671    False   
2  000071667F50BAFEA722A8E8284581E5   18    58    False  1378427564    False   
3  00007745A6EE969F1A0F44B10DC17671  268   526    False  1252294800    False   
4  00008C78BF1DDCC9FDCBAFFAF318A175   34    36    False  1538872824    False   

   Unnamed: 20  Unnamed: 21  Unnamed: 22   Unnamed: 23  
0          NaN          NaN          NaN           NaN  
1          NaN          NaN          NaN           NaN  
2          NaN          NaN          NaN  1.581305e+09  
3          NaN          NaN          NaN           NaN  
4          NaN          NaN          NaN           NaN

In [23]:
df = pd.read_csv(file, header=None, names=column_names, delimiter='\x01')

In [24]:
pd.set_option('display.max_columns', None)
print(df.shape)
display(df.head())

(80425, 24)


text_tokens  \
0  101\t56898\t137\t174\t63247\t10526\t131\t3197\...   
1  101\t102463\t10230\t10105\t21040\t10169\t12811...   
2  101\t56898\t137\t11255\t22037\t10263\t168\t111...   
3  101\t13073\t28757\t106\t100\t14120\t131\t120\t...   
4  101\t3460\t1923\t6632\t2824\t30368\t2179\t1881...   

                           hashtags                          tweet_id  \
0                               NaN  3C21DCFB8E3FEC1CB3D2BFB413A78220   
1                               NaN  3D87CC3655C276F1771752081423B405   
2  DB32BD91C2F1B37BE700F374A07FBC61  3701848B96AA740528A2B0E247777D7D   
3                               NaN  18176C6AD2871729384062F073CCE94D   
4                               NaN  AF11AF01F842E7F120667B7B0B38676D   

  present_media                     present_links  \
0         Video                               NaN   
1           NaN  BB422AA00380E45F312FD2CAA75F4960   
2           NaN  2423BA02A75DB2189335DDC3FB6B74A1   
3         Video                               NaN   
4           NaN                               NaN   

                    present_domains tweet_type  \
0                               NaN    Retweet   
1  92D397F8E0F1E77B36B8C612C2C51E23   TopLevel   
2  6D323BE93766E79BE423FAC5C28BE39B    Retweet   
3                               NaN   TopLevel   
4                               NaN      Quote   

                           language  tweet_timestamp  \
0  76B8A9C3013AE6414A3E6012413CDC3B       1581467323   
1  D3164C7FBCF2565DDF915B1B3AEFB1DC       1580975391   
2  22C448FF81263D4BAF2A176145EE9EAD       1581257232   
3  D3164C7FBCF2565DDF915B1B3AEFB1DC       1581164292   
4  22C448FF81263D4BAF2A176145EE9EAD       1581233650   

               engaged_with_user_id  engaged_with_user_follower_count  \
0  D1AA2C85FA644D64346EDD88470525F2                               737   
1  4DC65AC7BD963DE1F7617C047C33DE99                          52366425   
2  5C671539CB41B9807E209349B101E9FF                               988   
3  70B900BE17416923D1E236A38798F202                           1228134   
4  E94C0E9E8494F3D603F9D1A5C5242E3D                                73   

   engaged_with_user_following_count  engaged_with_user_is_verified  \
0                                706                          False   
1                               2383                           True   
2                                167                          False   
3                               5413                          False   
4                                299                          False   

   engaged_with_user_account_creation                  engaging_user_id  \
0                          1403069820  000046C8606F1C3F5A7296222C88084B   
1                          1230139136  00006047187D0D18598EF12A650E1DAC   
2                          1530094483  0000648BAA193AE4C625DDF789B57172   
3                          1378699943  000071667F50BAFEA722A8E8284581E5   
4                          1549054499  00007745A6EE969F1A0F44B10DC17671   

   engaging_user_follower_count  engaging_user_following_count  \
0                           131                           2105   
1                            22                             50   
2                           251                            719   
3                            18                             58   
4                           268                            526   

   engaging_user_is_verified  engaging_user_account_creation  \
0                      False                      1573978269   
1                      False                      1340673962   
2                      False                      1456473671   
3                      False                      1378427564   
4                      False                      1252294800   

   engaged_follows_engaging  reply_timestamp  retweet_timestamp  \
0                     False              NaN                NaN   
1                     False              NaN                NaN   
2       

Parse attributes containing tab-separated lists into lists. Transform raw timestamps into human-readable timestamps.

In [25]:
df['text_tokens'] = df['text_tokens'].str.split('\t')

def to_hex_list(x):
    output = str(x).split('\t')
#     output = [int(val, 16) for val in str(x).split('\t')] 
    return output

cols_to_process = ['hashtags', 'present_media', 'present_links', 'present_domains']

for col in cols_to_process:  
    df[col] = df[col].apply(lambda x: to_hex_list(x) if isinstance(x, str)  else x)


    
cols_to_process = ['tweet_timestamp', 'engaging_user_account_creation', 'reply_timestamp', 'retweet_timestamp', 'retweet_with_comment_timestamp', 'like_timestamp']

for col in cols_to_process:  
    df[col] = df[col].apply(lambda x: pd.Timestamp(x, unit='s'))

In [26]:
display(df.head())

text_tokens  \
0  [101, 56898, 137, 174, 63247, 10526, 131, 3197...   
1  [101, 102463, 10230, 10105, 21040, 10169, 1281...   
2  [101, 56898, 137, 11255, 22037, 10263, 168, 11...   
3  [101, 13073, 28757, 106, 100, 14120, 131, 120,...   
4  [101, 3460, 1923, 6632, 2824, 30368, 2179, 188...   

                             hashtags                          tweet_id  \
0                                 NaN  3C21DCFB8E3FEC1CB3D2BFB413A78220   
1                                 NaN  3D87CC3655C276F1771752081423B405   
2  [DB32BD91C2F1B37BE700F374A07FBC61]  3701848B96AA740528A2B0E247777D7D   
3                                 NaN  18176C6AD2871729384062F073CCE94D   
4                                 NaN  AF11AF01F842E7F120667B7B0B38676D   

  present_media                       present_links  \
0       [Video]                                 NaN   
1           NaN  [BB422AA00380E45F312FD2CAA75F4960]   
2           NaN  [2423BA02A75DB2189335DDC3FB6B74A1]   
3       [Video]                                 NaN   
4           NaN                                 NaN   

                      present_domains tweet_type  \
0                                 NaN    Retweet   
1  [92D397F8E0F1E77B36B8C612C2C51E23]   TopLevel   
2  [6D323BE93766E79BE423FAC5C28BE39B]    Retweet   
3                                 NaN   TopLevel   
4                                 NaN      Quote   

                           language     tweet_timestamp  \
0  76B8A9C3013AE6414A3E6012413CDC3B 2020-02-12 00:28:43   
1  D3164C7FBCF2565DDF915B1B3AEFB1DC 2020-02-06 07:49:51   
2  22C448FF81263D4BAF2A176145EE9EAD 2020-02-09 14:07:12   
3  D3164C7FBCF2565DDF915B1B3AEFB1DC 2020-02-08 12:18:12   
4  22C448FF81263D4BAF2A176145EE9EAD 2020-02-09 07:34:10   

               engaged_with_user_id  engaged_with_user_follower_count  \
0  D1AA2C85FA644D64346EDD88470525F2                               737   
1  4DC65AC7BD963DE1F7617C047C33DE99                          52366425   
2  5C671539CB41B9807E209349B101E9FF                               988   
3  70B900BE17416923D1E236A38798F202                           1228134   
4  E94C0E9E8494F3D603F9D1A5C5242E3D                                73   

   engaged_with_user_following_count  engaged_with_user_is_verified  \
0                                706                          False   
1                               2383                           True   
2                                167                          False   
3                               5413                          False   
4                                299                          False   

   engaged_with_user_account_creation                  engaging_user_id  \
0                          1403069820  000046C8606F1C3F5A7296222C88084B   
1                          1230139136  00006047187D0D18598EF12A650E1DAC   
2                          1530094483  0000648BAA193AE4C625DDF789B57172   
3                          1378699943  000071667F50BAFEA722A8E8284581E5   
4                          1549054499  00007745A6EE969F1A0F44B10DC17671   

   engaging_user_follower_count  engaging_user_following_count  \
0                           131                           2105   
1                            22                             50   
2                           251                            719   
3                            18                             58   
4                           268                            526   

   engaging_user_is_verified engaging_user_account_creation  \
0                      False            2019-11-17 08:11:09   
1                      False            2012-06-26 01:26:02   
2                      False            2016-02-26 08:01:11   
3                      False            2013-09-06 00:32:44   
4                      False            2009-09-07 03:40:00   

   engaged_follows_engaging reply_timestamp retweet_timestamp  \
0                     False             NaT               NaT   
1                     False    

# Extract Attributes

## Independent variables
The following new variable were defined in the dataframe:

| Name | Type | Description |
| --- | --- | --- |
| no_Photo | Integer | number of photo media in the tweet |
| has_media | bool | Indicates that is there any media type related to the tweet or not? (should be omitted later, it was just a helper for further extraction) |
| no_Photo | Integer | number of photo media in the tweet |
| no_Video | Integer | number of video media in the tweet|
| no_GIF | Integer | number of GIF media in the tweet |
| no_Other_media | Integer | (maybe it's impossible) if there is an unknown media type, it will be counted |
| no_Hashtags | Integer | Count the number of hashtags in each tweet |
| no_Domains | Integer | Count the number of domains in each tweet |
| no_Links | Integer | Count the number of links in each tweet |
| no_Words_In_Tweet | Integer | Count the number of words in the tweet |

## Labels
| Name | Type | Description |
| --- | --- | --- |
| is_like | bool | indicates that was there a like by the engaging user on the tweet (decided on that is there timestamp exists or not) |
| is_reply | bool | indicates that was there a reply by the engaging user on the tweet (decided on that is there timestamp exists or not) |
| is_retweet | bool | indicates that was there a retweet by the engaging user on the tweet (decided on that is there timestamp exists or not) |
| is_retweet_with_comment | bool | indicates that was there a retweet with comment by the engaging user on the tweet (decided on that is there timestamp exists or not) |

## Further Ideas to variable extraction:
Edit the following table if you want to propose new attribute

| Name | Type | Description |
| --- | --- | --- |
| attribute_name |attribute_type  | attribute_description |


In [27]:
df['present_media'].unique

<bound method Series.unique of 0        [Video]
1            NaN
2            NaN
3        [Video]
4            NaN
          ...   
80420        NaN
80421        NaN
80422        NaN
80423    [Photo]
80424        NaN
Name: present_media, Length: 80425, dtype: object>

In [28]:
pd.Series(df['present_media']).dropna().apply(tuple).unique()

array([('Video',), ('Photo',), ('Photo', 'Photo'),
       ('Photo', 'Photo', 'Photo'), ('GIF',),
       ('Photo', 'Photo', 'Photo', 'Photo')], dtype=object)

Is there any media in the cell?

In [29]:
df['has_media'] = df['present_media'].notna()

In [30]:
df[['has_media','present_media']]

has_media present_media
0           True       [Video]
1          False           NaN
2          False           NaN
3           True       [Video]
4          False           NaN
...          ...           ...
80420      False           NaN
80421      False           NaN
80422      False           NaN
80423       True       [Photo]
80424      False           NaN

[80425 rows x 2 columns]

df['no_Photo'] = df['no_Photo'].apply(lambda inp: countMedia(df['has_media'],  "Photo"), axis=1)
   #mediaPhoto(df['has_media'],df['present_media']))

df['no_Photo']

## Count the different media types for every tweet

In [31]:
photoCnt = []
videoCnt = []
gifCnt = []
other = []
for d,b in zip(df['present_media'],  df['has_media']):
    if b is True:
        pht = d.count("Photo")
        video = d.count("Video")
        gif  = d.count("GIF")
        photoCnt.append(pht)
        videoCnt.append(video)
        gifCnt.append(gif)        
        if len(d) != int(pht + video + gif):
             other.append(len(d) - int(pht + video + gif))
        else:
             other.append(0)
    else:
        photoCnt.append(0)
        videoCnt.append(0)
        gifCnt.append(0)
        other.append(0)                       

In [32]:
df['no_Photo'] = photoCnt
df['no_Video'] = videoCnt
df['no_GIF'] = gifCnt
df['no_Other_Media'] = other

df['no_Photos'] = df['no_Photo'].apply(lambda x: int(x['present_media'].str.count("Photo")) if df['has_media'].any() else 0)

In [33]:
df[['present_media','no_Photo', 'no_Video', 'no_GIF', 'no_Other_Media']].head(5)

present_media  no_Photo  no_Video  no_GIF  no_Other_Media
0       [Video]         0         1       0               0
1           NaN         0         0       0               0
2           NaN         0         0       0               0
3       [Video]         0         1       0               0
4           NaN         0         0       0               0

## Count the number of hashtags in each tweet

In [34]:
df['no_Hashtags'] = df.apply(lambda x : len(str(x['hashtags']).split()) if str(x['hashtags']) != "nan" else 0,axis=1)

In [35]:
#pd.set_option('display.max_colwidth', None)
df[['hashtags','no_Hashtags']].head(5)

hashtags  no_Hashtags
0                                 NaN            0
1                                 NaN            0
2  [DB32BD91C2F1B37BE700F374A07FBC61]            1
3                                 NaN            0
4                                 NaN            0

In [36]:
df[df['no_Hashtags'] != 0][['hashtags', 'no_Hashtags']].head(20)

hashtags  no_Hashtags
2                  [DB32BD91C2F1B37BE700F374A07FBC61]            1
10                 [9887C2F9C8FFECE3524054D91E871F84]            1
12                 [94838D747A18E1270C051B6DDFCAAE0D]            1
16                 [0EE6C006799F0AFD5655BDC4418B9A07]            1
17  [CA7AF9A82452F923905E4F8A485DED43, 9E53645466A...            6
19                 [DC3CA8E5056079E402ED35A7B1FDEC23]            1
20                 [E2774F85FC328CCF3515AF7B12EF18F0]            1
21                 [9A65AD335F99CB560602444D4F9111C8]            1
22  [1BBE947EB038868E02C649D57688A100, 7C209081942...            2
31  [CD33508ADA28A80D3A0DD6728CD7A537, 6F5D51E47F9...            2
34                 [B114D402D84DB8F5CCB91F25795F444E]            1
36                 [B6024D6F02AFAF9BC8ED941F72C228F1]            1
37  [F56A64A9F426D87BEA76E9A296EE006C, 47FB0777794...           10
38  [FE4B1491A2B93DE49B401E0D1279784E, FB5A802DF7D...            3
39  [4BDBA4AFD880EC2978072C923091EFB5, 174207A25D4...            3
40  [283E10F8BE4496B4BD7CEB62BD9A277C, ACF41558C81...            2
41  [1D49A2A8F6340E66BDD04A2E0516B260, 459A2A009D7...            3
47  [466397B084F7F0E479420BA91EC3F291, AD8BA848001...            4
48                 [16554C336F78240B4BA2B369D7EE1BAA]            1
49                 [EF48F0520EB04007A1864B892CB38AD7]            1

## Count the number of links in each tweet

In [37]:
df['no_Links'] = df.apply(lambda x : len(str(x['present_links']).split()) if str(x['present_links']) != "nan" else 0,axis=1)

In [38]:
df[df['no_Links'] != 0][['present_links', 'no_Links']].head(20)

present_links  no_Links
1                   [BB422AA00380E45F312FD2CAA75F4960]         1
2                   [2423BA02A75DB2189335DDC3FB6B74A1]         1
5                   [BB79CD318A68247B64F0E0BE7AFD5A92]         1
8                   [EF5D0A312E7A9BAEFBEA72A062E5F6CE]         1
12                  [F28E60037FB10D95E6CF3ADD83D67EA4]         1
19                  [AABF8A38DBBEE6D8B6FA3650262AC2D4]         1
31                  [F30118EB13346C80D754EAC2067092D5]         1
36                  [60DE00805B565526B32FF78F6EF0B9F7]         1
37                  [6EA7E07DDEDDBB1AAF7C7D9BFC232CAA]         1
43                  [9B1A416C5A7E6FED4A461F9F2108EB83]         1
49   [459F45BFE472BA192BC0615493A50BDB, 7ECBAA4765B...         2
64                  [1CB2E9D79689E547B32144E76BFA834F]         1
70                  [FFD41FE58DCA44A252A10E12DE9209CD]         1
75                  [0EDA872A7CD8EF9EF41AA3AF5C4DF38E]         1
79                  [DB7B100FFDA2EE43C3890CF6B7FD3908]         1
84                  [64437F367209412FA2B7DBA65A99716E]         1
89                  [6A287C68AA99FE049FDD579A70A86693]         1
95                  [94A6D5DAFD9DCAC96716D92A9FD4EF5B]         1
102                 [F8BBF1C5FCE5F850806B20E5191E87C7]         1
114                 [6B3A4D4CAA6B87A30A55AB9914C2B901]         1

In [39]:
df[['present_links', 'no_Links']].head(5)

present_links  no_Links
0                                 NaN         0
1  [BB422AA00380E45F312FD2CAA75F4960]         1
2  [2423BA02A75DB2189335DDC3FB6B74A1]         1
3                                 NaN         0
4                                 NaN         0

## Count the number of domains in each tweet

In [40]:
df['no_Domains'] = df.apply(lambda x : len(str(x['present_domains']).split()) if str(x['present_domains']) != "nan" else 0,axis=1)

In [41]:
df[df['no_Domains'] != 0][['present_domains', 'no_Domains']].head(20)

present_domains  no_Domains
1                   [92D397F8E0F1E77B36B8C612C2C51E23]           1
2                   [6D323BE93766E79BE423FAC5C28BE39B]           1
5                   [3896E26D12C903F0A00B6B1BE9A9BEA3]           1
8                   [3183ACF54B4022B25B4157B81C174DD5]           1
12                  [0F510BF067278AAD1ECAECC3380B0162]           1
19                  [3896E26D12C903F0A00B6B1BE9A9BEA3]           1
31                  [178B23A1119D4345DA847C10943997F9]           1
36                  [8F8CD97E2BFE3675BD1333C33D8F662B]           1
37                  [AA1AD16575EB76EB0CD1D2D951DC1F88]           1
43                  [7A52A6DD8470FC7770FDF9610909B635]           1
49   [3896E26D12C903F0A00B6B1BE9A9BEA3, 3896E26D12C...           2
64                  [F4A765D203E9604A20DF6C5362737718]           1
70                  [097C323141E21EA4941416B3BF4172E0]           1
75                  [1677479761CC167B998EF821EDC1331B]           1
79                  [B4A8AFACE5E163157B6128E79958C7E4]           1
84                  [8F261585DA0A3A006B3AFCC3368BE43A]           1
89                  [EC3B35626467D4D973FAA96D9101424E]           1
95                  [6825B061BF551A1617BE63A12C2A3CB6]           1
102                 [E85D1F68A0B4739BE39EAE4EEBB579D0]           1
114                 [1F5DE1B702AA81C85EB1D7B3D3029B50]           1

In [42]:
df[['present_domains', 'no_Domains']].head(5)

present_domains  no_Domains
0                                 NaN           0
1  [92D397F8E0F1E77B36B8C612C2C51E23]           1
2  [6D323BE93766E79BE423FAC5C28BE39B]           1
3                                 NaN           0
4                                 NaN           0

## Count the number of words in the tweet

It is also worth to mention, that there are separators in the text with ID [101], [131], [102] (but at least 2 separator in each tweet)

In [43]:
df['no_Words_In_Tweet'] = df.apply(lambda x : len(str(x['text_tokens']).split()) if str(x['text_tokens']) != "nan" else 0,axis=1)

In [44]:
df[["text_tokens",'no_Words_In_Tweet']].head(20)

text_tokens  no_Words_In_Tweet
0   [101, 56898, 137, 174, 63247, 10526, 131, 3197...                 31
1   [101, 102463, 10230, 10105, 21040, 10169, 1281...                 34
2   [101, 56898, 137, 11255, 22037, 10263, 168, 11...                 89
3   [101, 13073, 28757, 106, 100, 14120, 131, 120,...                 22
4   [101, 3460, 1923, 6632, 2824, 30368, 2179, 188...                 31
5   [101, 46242, 40751, 161, 100062, 10107, 10114,...                 45
6   [101, 56898, 137, 10192, 11373, 10500, 131, 21...                 13
7   [101, 65724, 15619, 22859, 14120, 131, 120, 12...                 20
8   [101, 100, 45031, 43804, 10121, 10146, 89387, ...                 59
9   [101, 56898, 137, 26037, 91678, 168, 15734, 73...                 47
10  [101, 10725, 14015, 14107, 13028, 10114, 12688...                 44
11  [101, 1925, 11662, 111801, 14244, 2452, 3709, ...                105
12  [101, 10117, 14329, 44871, 10107, 10301, 10893...                 63
13  [101, 35888, 10146, 74176, 55818, 12292, 10121...                 34
14  [101, 11005, 10116, 10106, 84732, 11272, 11904...                 24
15  [101, 11065, 21852, 117, 10549, 10944, 100, 18...                 62
16  [101, 56898, 137, 72337, 10858, 25096, 10115, ...                 46
17  [101, 4854, 4097, 1933, 6667, 1921, 89052, 225...                 87
18  [101, 4867, 1980, 8462, 100, 14120, 131, 120, ...                 22
19  [101, 60635, 25474, 10317, 10338, 10891, 11327...                 43

Average number of words

In [45]:
np.mean(df['no_Words_In_Tweet'])

48.516990985390116

Minimum number of words

In [46]:
np.min(df['no_Words_In_Tweet'])

2

Max number of words

In [47]:
np.max(df['no_Words_In_Tweet'])

255

## Encode what kind of interactions was the engaging user did with the tweet
Decided on timestamp: if there is timestamp, the interaction has happened (1), else no (0)
- Is_reply 
- Is_retweet
- is _retweet_with_comment
- Is_like

In [48]:
df['is_reply'] = df['reply_timestamp'].notna()
df['is_like'] = df['like_timestamp'].notna()
df['is_retweet_with_comment'] = df['retweet_with_comment_timestamp'].notna()
df['is_retweet'] = df['retweet_timestamp'].notna()

In [49]:
df[['reply_timestamp','is_reply', 'like_timestamp', 'is_like', 'retweet_with_comment_timestamp', 'is_retweet_with_comment','retweet_timestamp' ,'is_retweet']].head(20)

reply_timestamp  is_reply      like_timestamp  is_like  \
0                  NaT     False                 NaT    False   
1                  NaT     False                 NaT    False   
2                  NaT     False                 NaT    False   
3                  NaT     False 2020-02-10 03:29:24     True   
4                  NaT     False                 NaT    False   
5                  NaT     False                 NaT    False   
6                  NaT     False                 NaT    False   
7                  NaT     False 2020-02-12 22:54:31     True   
8                  NaT     False                 NaT    False   
9                  NaT     False 2020-02-12 22:49:15     True   
10                 NaT     False                 NaT    False   
11                 NaT     False 2020-02-07 15:19:38     True   
12                 NaT     False 2020-02-09 12:05:31     True   
13                 NaT     False                 NaT    False   
14                 NaT     False 2020-02-10 01:35:54     True   
15 2020-02-06 01:08:19      True                 NaT    False   
16                 NaT     False                 NaT    False   
17                 NaT     False 2020-02-10 09:51:37     True   
18                 NaT     False 2020-02-08 09:53:47     True   
19                 NaT     False                 NaT    False   

   retweet_with_comment_timestamp  is_retweet_with_comment  \
0                             NaT                    False   
1                             NaT                    False   
2                             NaT                    False   
3                             NaT                    False   
4                             NaT                    False   
5                             NaT                    False   
6                             NaT                    False   
7                             NaT                    False   
8                             NaT                    False   
9                             NaT                    False   
10                            NaT                    False   
11                            NaT                    False   
12                            NaT                    False   
13                            NaT                    False   
14                            NaT                    False   
15                            NaT                    False   
16                            NaT                    False   
17                            NaT                    False   
18                            NaT                    False   
19                            NaT                    False   

     retweet_timestamp  is_retweet  
0                  NaT       False  
1                  NaT       False  
2                  NaT       False  
3                  NaT       False  
4                  NaT       False  
5                  NaT       False  
6                  NaT       False  
7                  NaT       False  
8                  NaT       False  
9                  NaT       False  
10 2020-02-08 10:23:25        True  
11                 NaT       False  
12                 NaT       False  
13                 NaT       False  
14                 NaT       False  
15                 NaT       False  
16                 NaT       False  
17                 NaT       False  
18                 NaT       False  
19                 NaT       False